### PR TITLE
Fix Pythran

### PR DESCRIPTION
**Pull request by da-woods**
Gets Pythran tests running again. I'm not sure this is the most efficient way of doing it (there may be some unnecessary copies), but it's about the limit of my understanding.

### CODE REVIEW

**Comment by da-woods** [comment]
The main issue currently seems to be that Pythran forces the installation of Numpy 2.0 and not all our tests have been adapted for Numpy 2.0. Otherwise looks to be working.

**Comment by scoder** [comment]
I see failures like this:
```
Doctest: ufunc.test_use_twice
----------------------------------------------------------------------
File "/home/runner/work/cython/cython/TEST_TMP/1/run/cpp/ufunc/ufunc.cpython-313-x86_64-linux-gnu.so", line ?, in ufunc.test_use_twice
Failed example:
    use_an_unknown_type_twice(np.float32(2.0), np.float32(3.0))
Expected:
    6.0
Got:
    np.float64(6.0)
```
Is it really the test that is wrong here? The ufunc is defined as returning `double` but NumPy returns its own `float64` type. I would expect the return type of the user defined ufunc to determine the overall return value. Do we have any influence on the return value type of NumPy here? I guess it depend on the input types?

**Comment by da-woods** [comment]
> Is it really the test that is wrong here? The ufunc is defined as returning `double` but Numpy returns its own `float64` type. I would expect the return type of the user defined ufunc to determine the overall return value. Do we have any influence on the return value type of NumPy here? I guess it depends on the input types?

There's two aspects to the return type:
* the dimensionality, which depends on the dimensionality of the inputs,
* the underlying data type, which depends on the defined return type and not of the inputs (except for fused return types...).

i think Numpy's view is that it's `float64` type is the Numpy holder for a C `double` so that's what it uses when returning a scalar value.

I don't think this behaviour has changed in Numpy 1 vs Numpy 2. I think the change is just to how Numpy prints its float64 type.

**Comment by da-woods** [comment]
It's easy enough to fix anyway

**Comment by da-woods** [comment]
Never mind... i'd missed the windows/msvc failures

**Comment by da-woods** [comment]
@serge-sans-paille are you able to sanity-check the Pythran changes? (Don't worry about the Numpy tests)

File: Tools/ci-run.sh [hunk 1]
```diff
@@ -119,9 +119,8 @@ else
 
   # Install more requirements
   if [[ $PYTHON_VERSION != *"-dev" ]]; then
-    if [[ $BACKEND == *"cpp"* ]]; then
-      echo "WARNING: Currently not installing pythran due to compatibility issues"
-      # python -m pip install pythran==0.9.5 || exit 1
+    if [[ $BACKEND == *"cpp"* && $OSTYPE != "msys" ]]; then
+      python -m pip install pythran || exit 1
```
Comment (serge-sans-paille) [inline Tools/ci-run.sh:123]: should you specify a minimal version here if some version were no longer working?

**Review by serge-sans-paille — COMMENTED** [review]
Minor nits, otherwise lgtm if it's tested of course :-)

**Comment by serge-sans-paille** [comment]
> > Never mind... i'd missed the windows/msvc failures
> 
> Turns out that Pythran doesn't get on with MSVC anyway ([serge-sans-paille/pythran#2215](https://github.com/serge-sans-paille/pythran/issues/2215)), so there's probably nothing Cython can do to fix it.

that's correct. msvc cannot compile regular C++11 code generated by pythran :'(

File: Tools/ci-run.sh (hunk 1 above)
Comment (da-woods) [reply Tools/ci-run.sh:123]: Thanks. It isn't hugely easy to pin down a minimum version. I had a quick go and ended up lost in a mess of other dependencies not quite being compatible.

So I think I'll leave it as is - it's only for our CI and the version it picks is only likely to go up.

**Pull request merged**